Offset

jjimenezshaw opened the pull request:
I need to synchronize, not the centers of both maps, but other point. To create a continuous map, or a mosaic, or for any other purpose.

On this pull request the user can pass a function `offsetFn` that defines this transformation. It is very generic to allow flexibility. However, the most common use is really easy with the offsetHelper factory (as explained in the README.md).

Now, on each setView or similar function that resets a center, this center is computed with the provided function.

The default value for this option is an identity function, that is, returns the same center.

jieter commented on L.Map.Sync.js:
```diff
@@ -41,6 +48,27 @@
         },
 
+        // Helper function to compute the offset easily
+        // The arguments are relative positions respect refence and target maps
```
Sentence seems to miss some (parts of) words (added with **emphasis**):
  ... are relative positions **with** respect to ref**er**ense and target maps.

jjimenezshaw replied:
Thanks

jieter commented on L.Map.Sync.js:
```diff
@@ -41,6 +48,27 @@
+                var rs = refMap.getSize();
+                var ts = tgtMap.getSize();
+                var pt = refMap.project(center, zoom)
+                               .subtract([(0.5-or[0])*rs.x, (0.5-or[1])*rs.y])
```
Please conform to project style: use spaces around operators

jjimenezshaw replied:
Sure!
.eslint was not complaining on the tests.

jieter commented on README.md:
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
+
+As well `mapB.sync(mapA, {offsetFn: mapB.offsetHelper([0, 0], [1, 0.5])});` will sync the top left corner `[0 ,0]` in mapB with the center of the right side `[1, 0.5]` in mapA.
+
+If you want the actions to be synced vice-versa, you should use simetric values (as reference and target are swapped).
```
simetric -> symmetric.

I think this line could use an example of a non-center, symmetric sync pair.

jieter commented on README.md:
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
+
+If you want the actions to be synced vice-versa, you should use simetric values (as reference and target are swapped).
+
+The default behaviour uses `[0.5, 0.5], [0.5, 0.5]`, that synchronizes the centers.
```
I would rephrase:

The default behaviour is to synchronize the centers, the corresponding offset points are `[0.5, 0.5], [0.5, 0.5]`.

or something like that.

jieter commented on L.Map.Sync.js:
```diff
@@ -41,6 +48,27 @@
+        // with the top right corner of the target map.
+        // The values can be less than 0 or greater than 1. It will sync
+        // points out of the map.
+        offsetHelper: function (ratioRef, ratioTgt) {
```
Not sure if this method should be at home in this location, I mean, it doesn't interact with the instance at all. Is `L.Util` a better location?

jjimenezshaw replied:
While i was coding it I was thinking on where to place it. Yes, it is an "static" function (in Java or C++ notation), or a functions factory. It does not interact with the instance at all. That lets you more freedom to use your own function if you want something fancier.
However, it is going to be used only in this context, to create the function expected in offsetFn.

Maybe makes sense to move it to `L.Util.offsetHelper`... or with a different name?

jieter commented on README.md:
```diff
@@ -36,6 +36,21 @@ mapB.sync(mapC);
+The parameters send to the function are `(center, zoom, referenceMap, targetMap)`, and it must return the equivalent center to produce your offset. That means, the center to pass to setView.
+
+In most cases, you can use the factory `offsetHelper`, that accepts two arrays of two elements each `(ratioRef, ratioTgt)`. The meaning of this array is the relative position -relative to the top left corner and the whole size- in the map container of the point to synchronize. (1 is for the whole width or height). Values greater than 1 or less than 0 work fine.
+
```
From this paragraph, I wonder:
 - can you add/is it beneficial to have a use case + example of a sync scenario which is not possible by using the `offsetHelper`.
- "Values greater than 1 or less than 0 work fine." -> might be  good to explain this a little more/give an example with the two points + what they mean.

jieter commented:
Interesting feature, thanks for the effort. I've added some comments and will have a look to the actual implementation later.
Might be interesting to record some GIFs and add them to the documentation.

jjimenezshaw commented:
@jieter Thanks for your comments. I am open to new ways to explain the feature in the README file. Took me a long time writing it... but still I am not happy with the result. As I know how it works, I understand it. But I am sure it will be difficult for a new user. Feel free to suggest any change ;)

jjimenezshaw commented:
@jieter When you say 

> Might be interesting to record some GIFs and add them to the documentation

do you mean an animated GIF on the README.md?

jieter commented:
Yes, (small) animated GIFs in the readme.

jjimenezshaw commented:
@jieter About different functions, I have just created one that syncs the maps in a global view. The maps will be displayed as a continuous one, doesn't matter where are the divs. If they overlap, they fit perfectly.
```JavaScript
function offsetGlobal(center, zoom, refMap, tgtMap) {                           
    var refC = refMap.getContainer();                                           
    var tgtC = tgtMap.getContainer();                                           
    var pt = refMap.project(center, zoom)                                       
                   .subtract([refC.offsetLeft, refC.offsetTop])                 
                   .subtract(refMap.getSize().divideBy(2))                       
                   .add([tgtC.offsetLeft, tgtC.offsetTop])                      
                   .add(tgtMap.getSize().divideBy(2));                          
    return refMap.unproject(pt, zoom);                                          
}        
```
Should I include it as part of the code (in `L.Util`) or just as an example in the README.md?

jjimenezshaw commented:
@jieter I fixed all your comments (I hope), and added the animated gif (it is a bit small, I don't want to include a 2MB file just for that).

Apart from including or not `offsetGlobal` function in the code, I am done.

jieter commented:
I'd say add it to the docs, not to the code. I'll merge this, you can open a new PR adding it to the docs.

jieter commented:
0.1.0 is the latest of 6 releases
